### PR TITLE
Improve job.data, project.data (H5Store) examples.

### PR DESCRIPTION
**Pull request by bdice**
## Description
I created this PR to start addressing this issue: https://github.com/glotzerlab/signac-docs/issues/50

@klywang Do you have specific suggestions on how to improve this? (I don't think I've "solved" the issue yet, since I haven't provided explicit examples for `job.data` as requested.) Feel free to edit this PR directly.

## Motivation and Context
https://github.com/glotzerlab/signac-docs/issues/50

## Types of Changes

- [x] Documentation update
- [ ] Bug fix
- [ ] New feature
- [ ] Breaking change<sup>1</sup>

<sup>1</sup>The change breaks (or has the potential to break) existing functionality.

## Checklist:

- [x] I am familiar with the [**Contributing Guidelines**](https://github.com/glotzerlab/signac/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**Contributor Agreement**](https://github.com/glotzerlab/signac/blob/master/ContributorAgreement.md).
- [ ] My name is on the [list of contributors](https://github.com/glotzerlab/signac/blob/master/contributors.yaml).
- [ ] My code follows the [code style guideline](https://github.com/glotzerlab/signac/blob/master/CONTRIBUTING.md#code-style) of this project.
- [ ] The changes introduced by this pull request are covered by existing or newly introduced tests.

If necessary:
- [ ] I have updated the API documentation as part of the package doc-strings.
- [ ] I have created a separate pull request to update the [framework documentation](https://docs.signac.io/) on [signac-docs](https://github.com/glotzerlab/signac-docs) and linked it here.
- [ ] I have updated the [changelog](https://github.com/glotzerlab/signac/blob/master/changelog.txt).

### CODE REVIEW

**Comment by vyasr** [comment]
@klywang any feedback on this PR?

**Comment by csadorf** [comment]
Why is this PR not assigned to anybody?

**Comment by vyasr** [comment]
This PR was in draft mode until recently. When moved out of draft mode, what is our policy for assigning a maintainer in charge? I'm guessing the bot didn't add any reviewers, so nobody assigned themself to the issue.

**Comment by bdice** [comment]
I assigned myself to this PR and requested review from @klywang since she originally reported the documentation challenges. I marked it as "ready for review" (no longer draft mode) in hopes that it would get reviewed by someone, but I suppose the bot didn't assign reviewers.

**Comment by klywang** [comment]
Is there a specific reason you one should use `H5Store` over `job.data` or `project.data`? If there isn't a good reason, it would help to have an short example that shows its usage with the context manager. For example:
>with job.data
>    net_energy = job.data.net_energy[:]
Just enough so that someone who wants to use the functionality can glance at the docs and know how to use it.

**Comment by bdice** [comment]
@klywang The purpose of linking to `H5Store` is that the `job.data` _is_ an `H5Store`. I was hesitant to duplicate examples for `job.data` and `project.data`, just to keep the code organized (changes to `H5Store` would have to be reflected in several examples, not just `job.data` and `project.data` but also `job.stores`, etc. -- hard to know exactly where to draw the line). That said, I see why it would be user-friendly to have a snippet like that in the docs. I'll adapt your example and ping you for review.

**Comment by csadorf** [comment]
> @klywang The purpose of linking to `H5Store` is that the `job.data` _is_ an `H5Store`.

Also might be worthwhile to try to make that relationship more explicit in the documentation to avoid confusion.

**Comment by klywang** [comment]
@csadorf 
Poor wording on my end. I meant, under what circumstances should a user use `H5Store` or `job.data`\ `project.data`? I thought the documentation clearly communicated the fact that these features were `H5Stores`. The primary source of confusion was that there were a lot of different ways to do a similar operation and it was difficult to understand which version I should use, especially before I knew how to use it.

**Comment by csadorf** [comment]
> @csadorf
> Poor wording on my end. I meant, under what circumstances should a user use `H5Store` or `job.data`\ `project.data`? I thought the documentation clearly communicated the fact that these features were `H5Stores`. The primary source of confusion was that there were a lot of different ways to do a similar operation and it was difficult to understand which version I should use, especially before I knew how to use it.

@klywang I see. So it's more about when to use `job.data`, `project.data`, or a direct instance of `H5Store`.  Could you maybe propose a paragraph that would have helped you to understand that better/earlier?

**Comment by bdice** [comment]
(I would also include `job.stores` in that description if possible)

**Comment by klywang** [comment]
**For `job.data`**
`job.data` lets you read and write data associated with your `job`. It should be used for large array-like data, which can't be stored efficiently in the job document. For examples and usage, see [link for topic guide].



**For Topic Guides > Jobs > Job Data Store**
*Some revised examples for the Topic Guide already exist in this section, but I rewrote them to streamline the rest of these examples.*


*Writing data*
To write data to a store:
```python
job.data['velocities'] = np.ones([10, 3, 2])
```

Data may also be written in subgroups:
```python
job.data['net_energy/frame_10'] = np.ones([10, 3])
```

*Reading data*
To access data as an attribute:
```python
with job.data:
    velocities = job.data.velocities[:]
    net_energy = job.data.net_energy.frame_10[:]
```

To access data as a key:
```python
with job.data:
    velocities = job.data['velocities'][:]
    net_energy = job.data['net_energy/frame_10'][:]
```

*Other uses*
To view keys associated with `job.data` or in a subgroup:
```python
print(list(job.data.keys())
[Output]: ['velocities', 'net_energy']

print(list(job.data.net_energy.keys()))
[Output]: ['frame_10']
```

To iterate over keys in `job.data` or a subgroup:
```python
for frame_n in job.data.net_energy.keys():
    with job.data:
        frame_energy = job.data.net_energy[frame_n][:]
```

Data written with `job.data` is stored in a file named `signac_data.h5` in the associated `job` folder. For cases where `job`-associated data may be accessed from multiple sources at the same time or other instances where multiple files may be preferred to one large file, `job.stores` should be used instead of `job.data`.

For more information about supported data types and the file format, please see documentation for the `H5Store`.

Also, note for myself -- there is a typo in the `H5Store` doc that I will fix.
P.S. Thank you @vyasr @bdice for the input!

**Comment by vyasr** [comment]
@atravitz @b-butler I've requested some additional reviews on here, please read through @klywang's suggested improvements to our documentation in her most recent comment on this thread and provide some commentary. Thanks!

**Comment by atravitz** [comment]
As someone who has only just recently started using `job.data`, this is incredibly clear and helpful! One minor thing: it might be better to name `frame_10` something else. The way it currently reads is that it's an alternate way of accessing a frame, rather than being a sub-key.

**Comment by vyasr** [comment]
@klywang could you incorporate @atravitz's suggestion and then add your snippet to the relevant places in the documentation (i.e. update topic guides, then add links in the docstrings to point to the topic guides instead of just to the `H5StoreManager` and related classes).

**Comment by bdice** [comment]
@klywang We'll be making a release on Friday. If you have a chance to make the documentation changes by then (we also have a sprint on Thursday), that would be great!

**Comment by bdice** [comment]
It doesn't look like @klywang (or anyone) pushed any commits with the drafted text that is in [her comment above](https://github.com/glotzerlab/signac/pull/235#issuecomment-556454901). I'm confused about whether @b-butler and @atravitz meant to approve the file changes _as they are currently_ (which is incomplete), or approve the _concept of @klywang's changes_. The changes @klywang suggested look great to me, but we need to ...actually implement them in the docs. 😕 

I'll be making a release of signac tomorrow, **could someone take that on and copy-paste the snippets into the right places in the docs**? If so, please reply on this PR. Otherwise I will try to apply the changes myself before making the release. It seems like we're generally in agreement over their intended content.

**Comment by atravitz** [comment]
I’ll take care of this before 11 am tomorrow (Friday).




----------
On Thu, Dec 12, 2019 at 10:32 PM, Bradley Dice <notifications@github.com>
wrote:

It doesn't look like @klywang <https://github.com/klywang> (or anyone)
pushed any commits with the drafted text that is in her comment above
<https://github.com/glotzerlab/signac/pull/235#issuecomment-556454901>. I'm
confused about whether @b-butler <https://github.com/b-butler> and @atravitz
<https://github.com/atravitz> meant to approve the file changes *as they
are currently* (which is incomplete), or approve the *concept of @klywang
<https://github.com/klywang>'s changes*. The changes @klywang
<https://github.com/klywang> suggested look great to me, but we need to
...actually implement them in the docs. 😕

I'll be making a release of signac tomorrow, *could someone take that on
and copy-paste the snippets into the right places in the docs*? If so,
please reply on this PR. Otherwise I will try to apply the changes myself
before making the release. It seems like we're generally in agreement over
their intended content.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/glotzerlab/signac/pull/235?email_source=notifications&email_token=AHT6IXZT7O3ZN7GMOF7PFMTQYL657A5CNFSM4JE5YSUKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEGYZJ6Y#issuecomment-565286139>,
or unsubscribe

**Comment by b-butler** [comment]
@bdice I meant to approve the suggestion. I should have been more clear about that. Otherwise, like you said the PR is incomplete.

**Comment by atravitz** [comment]
Update: @klywang is adding these in, then I'll review

**Comment by vyasr** [comment]
Thanks for the update @atravitz!

File: signac/contrib/job.py [hunk 1]
```diff
@@ -308,9 +308,11 @@ def stores(self):
...
     @property
     def data(self):
-        """The data store associated with this job.
-
-        See :class:`~..core.h5store.H5Store` for usage examples.
+        """
+        Use to read and write data associated with a `job`. 
+        It should be used for large array-like data, which can't be stored 
+        efficiently in the job document. For examples and usage, see 
+        Topic Guides > Jobs > Job Data Storage.
```
Comment (atravitz) [inline signac/contrib/job.py]:
```suggestion
        The data associated with this job.
        This property should be used for large array-like data, which can't be stored 
        efficiently in the job document. For examples and usage, see 
        Topic Guides > Jobs > Job Data Storage.
```

File: signac/contrib/project.py [hunk 2]
```diff
@@ -336,7 +336,10 @@ def stores(self):
 
     @property
     def data(self):
-        """The data associated with this project.
+        """Use to read and write data associated with a `job`. It 
+        should be used for large array-like data, which can't be stored efficiently
```
Comment (atravitz) [inline signac/contrib/project.py]: To keep consistent with other doc strings:
```suggestion
"""The data associated with this project. 
This property should be used to read ...
```

**Comment by klywang** [comment]
Topic guide updates are [here](https://github.com/glotzerlab/signac-docs/pull/62)

File: signac/contrib/job.py (hunk 1 above)
Comment (klywang) [reply signac/contrib/job.py]: Resolved cc1eda

**Pull request merged**